### PR TITLE
Split call to check permissions into chunks

### PR DESCRIPTION
**Pull request by csrwng**
The call to check available permissions only allows 100 permissions at a
time per check. This change splits the call into multiple calls when
there are more than 100 permissions to check.

### CODE REVIEW

**Comment by csrwng** [comment]
@dgoodwin ptal

**Comment by dgoodwin** [comment]
Is it feasible to gets some unit tests around this? It looks like a good spot for an edge case  bug to slip in.

**Comment by csrwng** [comment]
> Is it feasible to gets some unit tests around this? It looks like a good spot for an edge case bug to slip in.

Yup, let me look into adding a test

File: pkg/controller/utils/gcp/utils.go [hunk 1]
```diff
@@ -151,18 +151,34 @@ func CheckPermissionsAgainstPermissionList(gcpClient ccgcp.Client, permList []st
...
+		permRequest := &cloudresourcemanager.TestIamPermissionsRequest{
+			Permissions: filteredPermList[index:upperLimit],
+		}
+		permResponse, err := gcpClient.TestIamPermissions(projectName, permRequest)
+		if err != nil {
+			return false, fmt.Errorf("error testing permissions: %v", err)
+		}
+		responses = append(responses, permResponse)
+		index += 100
 	}
```
Comment (ironcladlou) [inline pkg/controller/utils/gcp/utils.go]: Just another possible approach for consideration since I found the problem interesting:

```go
var chunks [][]string
chunkSize := 100
permLen := len(filteredPermList)
for i := 0; i < permLen; i += chunkSize {
    end := i + chunkSize
    if end > permLen {
        end = permLen
    }
    chunks = append(chunks, filteredPermList[i:end])
}

for chunk := range chunks {
  // do request
}
```

Comment (csrwng) [reply pkg/controller/utils/gcp/utils.go]: Thx! I used it for the test as an alternative way to come up with the chunks

**Comment by csrwng** [comment]
@dgoodwin added a unit test

File: pkg/controller/utils/gcp/utils.go [hunk 2]
```diff
@@ -147,22 +147,43 @@ func filterOutPermissions(permList, filterOutList []string) []string {
...
-	}
-	permResponse, err := gcpClient.TestIamPermissions(projectName, permRequest)
-	if err != nil {
-		return false, fmt.Errorf("error testing permissions: %v", err)
+	// Split list to only check 100 permissions at a time
+	responses := []*cloudresourcemanager.TestIamPermissionsResponse{}
+	done := false
+	index := 0
+	for !done {
+		upperLimit := len(filteredPermList)
```
Comment (wking) [inline pkg/controller/utils/gcp/utils.go]: nit: I'd rather use @ironcladlou's [loop approach][1] here and just hard-code the expected chunking in the unit test.

```go
for _, testCase := range []struct{
  length int
  chunks []int
}{
  {length: 0},
  {length: 2, chunks: {2}},
  {length: 100, chunks: {100}},
  {length: 101, chunks: {100, 1}},
  {length: 230, chunks: {100, 100, 30}},
  {length: 300, chunks: {100, 100, 100}},
  {length: 379, chunks: {100, 100, 100, 79}},
} {
  t.Run(fmt.Sprintf("length=%d", testCase.length), ...)
}
```

[1]: https://github.com/openshift/cloud-credential-operator/pull/94#discussion_r307866955

File: pkg/controller/utils/gcp/utils.go [hunk 3]
```diff
@@ -147,22 +147,43 @@ func filterOutPermissions(permList, filterOutList []string) []string {
...
+		}
+		responses = append(responses, permResponse)
+		index += 100
 	}
 
 	// check that each perm in our list is actually available
 	permMap := map[string]bool{}
-	for _, perm := range permResponse.Permissions {
-		permMap[perm] = true
+	for _, resp := range responses {
```
Comment (wking) [inline pkg/controller/utils/gcp/utils.go]: nit: instead of collecting a slice of responses, I'd rather create `permMap` before the chunk loop and then iterate over each chunk's response inside the loop:

```go
permMap := map[string]bool{}
for i := 0; i < permLen; i += chunkSize {
  ...
  for _, perm := range resp.Permissions {
    permMap[perm] = true
  }
}
```

**Comment by wking** [comment]
Left two minor nits inline, but this looks fine to me whether those get addressed or not.

**Comment by csrwng** [comment]
> Left two minor nits inline, but this looks fine to me whether those get addressed or not.

Thx! addressed.

File: pkg/controller/utils/gcp/utils.go [hunk 4]
```diff
@@ -147,37 +147,47 @@ func filterOutPermissions(permList, filterOutList []string) []string {
...
+	// Split list to only check 100 permissions at a time
+	var chunks [][]string
+	chunkSize := 100
+	permLen := len(filteredPermList)
+	for i := 0; i < permLen; i += chunkSize {
+		end := i + chunkSize
+		if end > permLen {
+			end = permLen
+		}
+		chunks = append(chunks, filteredPermList[i:end])
```
Comment (wking) [inline pkg/controller/utils/gcp/utils.go]: You don't need to build a `chunks` slice; just make your request with `Permissions: filteredPermList[I:end]` in this same loop.

Comment (csrwng) [reply pkg/controller/utils/gcp/utils.go]: Ack, fixed
